doc/developer/platform/formalism.md

```diff
@@ -0,0 +1,176 @@
+We call this pair of set of updates and `(since, upper)` frontiers a *partial TVC* or pTVC.
+
+## Evolution
+
+Materialize maintains a (partial) map from `GlobalId`s to pTVCs.
+A `GlobalId` is never re-used to mean a different TVC.
+
+The life cycle of each `GlobalId` follows several steps:
+
+0. The `GlobalId` is initially not yet bound.
+
+    In this state, the `GlobalId` cannot be used, as we do not yet know what its initial `since` will be.
+
+1. The `GlobalId` is bound to some definition, with an initially equal `since` and `upper`.
+
+    This means that the collection is initially not useful, as it cannot be queried at any time greater or equal to `since` but not greater or equal to `upper`.
+    However, the initial value of `since` indicates that one should expect the data to eventually become available for each time greater or equal to `since`.
+    The partial TVC can now be used in queries, though results may not be immediately available.
+
+2. Repeatedly, either of the these two steps occur:
+
+    a. The `upper` frontier advances, revealing more of the future of the pTVC.
+
+    b. The `since` frontier advances, concealing some of the past of the pTVC.
+
+3. Eventually, the `since` frontier advances to the empty collection, rendering the TVC complete and closed.
+
+The `GlobalId` is never reused.
+
+# Storage and Compute
+
+Materialize produces and maintains time-varying collections in two distinct ways:
+
+*   The **Storage** layer records explicit TVCs in response to direct instruction (e.g. user DML), integrations with external sources of data, and feedback from Materialize itself.
+    The representation in the Storage layer is as explicit update triples and maintained `since` and `upper` frontiers.
+    The primary property of the Storage layer is that, even across failures, it is able to reproduce all updates for times that lie between `since` and `upper` frontiers that it has advertised.
+
+*   The  **Compute** layer uses a view definition to transform input TVCs into the exactly corresponding output TVCs.
+    The output TVC varies *exactly* with the input TVCs:
+    For each `time`, the output collection at `time` equals the view applied to the input collections at `time`.
+
+The Storage and Compute layers both track `since` and `upper` frontiers for the collections they manage.
+For both layers, the `since` frontier is controlled by their users and it does not advance unless it is allowed.
+The `upper` frontier in Storage is advanced when it has sealed a frontier and made the data up to that point durable.
+The `upper` frontier in Compute is advanced when its inputs have advanced and its computation has pushed updates fully through the view definitions.
+
+## Capabilities
+
+The `since` and `upper` frontiers are allowed to advance through the use of "capabilities".
+
+* `ReadCapability(id, frontier)` prevents the `since` frontier associated with `id` from advancing beyond `frontier`.
```

Comment:
Is the `id` here the same as the `GlobalId` from above?

Reply:
That's the intent yes. Sounds like it wasn't clear, so I will elaborate in the text!

Reply:
Maybe just annotate the struct with types?

```
ReadCapability(id: GlobalId, at: Frontier)
```